sklearn/decomposition/online_lda.py

```diff
@@ -0,0 +1,668 @@
+        X : sparse matrix, shape=(n_samples, n_features)
+            Document word matrix.
+
+        Returns
+        -------
+        self
+        """
+
+        X = self._to_csr(X, "LatentDirichletAllocation.fit")
+        n_samples, n_features = X.shape
+        max_iter = self.max_iter
+        total_samples = n_samples * max_iter
+        evaluate_every = self.evaluate_every
+        learning_method = self.learning_method
+        batch_size = self.batch_size
+
+        # initialize parameters
+        self._init_latent_vars(n_features)
+        # change to perplexity later
+        last_bound = None
+        for i in xrange(max_iter):
+            if learning_method == 'online':
+                for idx_slice in gen_batches(n_samples, batch_size):
+                    self._em_step(X[idx_slice, :], total_samples=total_samples, batch_update=False)
+            else:
+                # batch update
+                doc_topics = self._em_step(X, total_samples=total_samples, batch_update=False)
+
+            # check perplexity
+            if evaluate_every > 0 and (i + 1) % evaluate_every == 0:
+                if learning_method == 'online':
+                    doc_topics, _ = self._e_step(X, False)
+
+                bound = self.perplexity(X, doc_topics, sub_sampling=False)
+                if self.verbose:
+                    print('iteration: %d, perplexity: %.4f' % (i + 1, bound))
+
+                if last_bound and abs(last_bound - bound) < self.perp_tol:
+                    break
+                last_bound = bound
+
+        return self
+
+    def transform(self, X):
+        """
+        Transform data X according to the fitted model.
+
+        Parameters
+        ----------
+        X : sparse matrix, shape=(n_samples, n_features)
+            Document word matrix.
+            `n_features` must be the same as `self.n_features`
+
+        Returns
+        -------
+        doc_topic_distr : array, [n_samples, n_topics]
+            Document topic distribution for X.
+        """
+
+        if not hasattr(self, 'components_'):
+            raise NotFittedError(
+                "no 'components_' attribute in model. Please fit model first.")
+
+        # make sure word size is the same in fitted model and new doc
+        # matrix
+        X = self._to_csr(X, "LatentDirichletAllocation.transform")
+        n_samples, n_features = X.shape
+        if n_features != self.n_features:
+            raise ValueError(
+                "feature dimension(vocabulary size) does not match.")
+
+        doc_topic_distr, _ = self._e_step(X, False)
+
+        if self.normalize_doc:
+            doc_topic_distr /= doc_topic_distr.sum(axis=1)[:, np.newaxis]
+        return doc_topic_distr
+
+    def _approx_bound(self, X, doc_topic_distr, sub_sampling):
+        """
+        Calculate approximate bound for data X and document topic distribution
+        Since log-likelihood cannot be computed directly, we use this bound
+        estimate it.
```

Comment:
The original implementation says "Estimates the variational bound over *all documents* using only the documents passed in as "docs."" which sounds different from what it says here.